src/go/goblins/docker/docker.go

```diff
@@ -0,0 +1,209 @@
+		storagePath := filepath.Join(rootDir, storageIhp)
+		volumeMountpoint := filepath.Join(mountDir, mountpointIhp)
+		args := []string{"--bind", "-oro", storagePath, volumeMountpoint}
+		commands = append(commands, exec.Command("mount", args...))
+	}
+	return commands
+}
+
+func mountVolumesDocker2(mountDir string, mountPoints []storageInfo) []*exec.Cmd {
+	commands := make([]*exec.Cmd, 0)
+	for _, s := range mountPoints {
+		srcMount := strings.TrimLeft(s.source, string(os.PathSeparator))
+		dstMount := strings.TrimLeft(s.destination, string(os.PathSeparator))
+		if srcMount == "" {
+			volumeName := s.name
+			srcMount = filepath.Join("docker", "volumes", volumeName, "_data")
+		}
+		storagePath := filepath.Join(mountDir, dstMount)
+		volumeMountpoint := filepath.Join(mountDir, dstMount)
+		args := []string{"--bind", "-oro", storagePath, volumeMountpoint}
+		commands = append(commands, exec.Command("mount", args...))
+	}
+	return commands
+}
+
+func overlayFsMountCommands(mountDir string, mountID string, dockerDir string,
+	storageMethod string, dockerVersion int) ([]*exec.Cmd, error) {
+	mountIDPath := filepath.Join(dockerDir, storageMethod, mountID)
+	const lowerdirName = "lower"
+	const upperdirName = "diff"
+
+	lowerDirPath := filepath.Join(mountIDPath, lowerdirName)
+	lowerContent, err := ioutil.ReadFile(lowerDirPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read lower content pointer: %s", err)
+	}
+
+	// There are differences between overlay and overlay2 here for the lower directory
+	if storageMethod == "overlay" {
+		lowerDirPath = filepath.Join(dockerDir, storageMethod, string(lowerContent), "root")
+	} else {
+		// For the overlay2 driver, the pointer to the 'lower directory' can be
+		// made of multiple layers.
+		// For that argument to be passed to the mount.overlay command, we need to
+		// reconstruct full paths to all these layers.
+		// ie: from 'abcd:0123' to '/var/lib/docker/abcd:/var/lib/docker/0123'
+		var paths []string
+		for _, lc := range strings.Split(string(lowerContent), ":") {
+			paths = append(paths, filepath.Join(dockerDir, storageMethod, lc))
+		}
+		lowerDirPath = strings.Join(paths, ":")
+	}
+	upperDirPath := filepath.Join(mountIDPath, upperdirName)
+
+	opts := fmt.Sprintf(`-oro,lowerdir=%s:%s`, upperDirPath, lowerDirPath)
+	args := []string{"-toverlay", opts, "overlay", mountDir}
+	command := exec.Command("mount", args...)
+	return []*exec.Cmd{command}, nil
+}
+
+// Umount a path that has been previously mounted.
+// Note that since some of the filesystems mount
+// over and over the same path, we need to keep umount!
+func Umount(path string) error {
+	cmd := exec.Command("umount", path)
+	err := cmd.Run()
+	log.Printf("umount %s", path)
+	// The first time we should always umount.
+	if err != nil {
+		return err
+	}
+	// Now keep umount till we hit an error (i.e. we are done)
+	for exec.Command("umount", path).Run() == nil {
+		log.Printf("still umounting %s", path)
```

Comment:
Nit: What if umount starts failing fast? This can get noisy. Maybe print a line before and after this loop.

Reply:
Fair.